lib/memjs/memjs.js

```diff
@@ -1,23 +1,40 @@
 // # MemJS Memcache Client
 
+var HashRing = require('hashring');
+
 var errors = require('./protocol').errors;
 var Server = require('./server').Server;
 var noopSerializer = require('./noop-serializer').noopSerializer;
 var makeRequestBuffer = require('./utils').makeRequestBuffer;
-var hashCode = require('./utils').hashCode;
+var addToRequestBuffer = require('./utils').addToRequestBuffer;
 var merge = require('./utils').merge;
 var makeExpiration = require('./utils').makeExpiration;
 var makeAmountInitialAndExpiration = require('./utils').makeAmountInitialAndExpiration;
+var constants = require('./constants');
 
 // Client initializer takes a list of `Server`s and an `options` dictionary.
 // See `Client.create` for details.
 var Client = function(servers, options) {
   this.servers = servers;
   this.seq = 0;
   this.options = merge(options || {},
-    {failoverTime: 60, retries: 2, retry_delay: 0.2, expires: 0, logger: console});
+    {retries: 2, retry_delay: 0.2, expires: 0, logger: console});
 
   this.serializer = this.options.serializer || noopSerializer;
+
+  this.serverMap = {};
+  for (var serverIndex in servers) {
+    var server = servers[serverIndex];
+    this.serverMap[server.host] = server;
+  }
+
+  this.hashring = new HashRing(
+    Object.keys(this.serverMap),
+    this.options.hashringAlgorithm || 'md5',
+    this.options.hashringOptions || {
+      compatibility: 'ketama'
+    }
+  );
```

Comment:
What do you think about making hashing function pluggable, with the default falling back to the existing implementation? This lets the library remain "zero dependency" and would make it easier to upstream.

It would also keep the definition of our hashing func in the main Notion codebase, which would make us more resilient to upstream changes and give us more options if we want to change key distribution again in the future. 

Not sure what the signature would be;  possibilities include `(key) => serverHostNameString`, `(key, serverHostNameArray) => serverHostNameString`, or `(key, bucketsNumber) => bucketNumber`

Reply:
going to try this today, and yeah, I'm not 100% sure what I want the signature to be, I will feel it out in my fingers

Reply:
went with:
//                            (servers: string[], key: string): string